tests/si/rendering.cc

```diff
@@ -49,30 +121,324 @@ APP("ui rendering")
+        ui_merger.viewport = {{0, 0}, {float(window.getWidth()), float(window.getHeight())}};
+        ui_merger.mouse_pos = tg::pos2(mouse);
+        ui_merger.is_lmb_down = input.get(action_mouse_left).isActive();
+        ui_merger.set_record_timings(std::chrono::duration<double>(t1 - t0).count());
+        ui.update(r, ui_merger);
+
+        // upload ui data
+        auto ui_proj = ctx.get_upload_buffer(sizeof(tg::mat4));
+        tg::mat4 proj = tg::translation(-1.f, 1.f, 0.0f) * tg::scaling(2.f / window.getWidth(), -2.f / window.getHeight(), 1.f);
+        ctx.write_to_buffer(ui_proj, proj);
+
+        size_t vertex_byte_size = 0;
+        size_t index_byte_size = 0;
+        for (auto const& rl : ui_merger.get_render_data().lists)
+        {
+            vertex_byte_size += rl.vertices.size_bytes();
+            index_byte_size += rl.indices.size_bytes();
+        }
+        auto ui_vertex_buffer = ctx.get_upload_buffer(vertex_byte_size);
+        auto ui_index_buffer = ctx.get_upload_buffer(index_byte_size, sizeof(int));
+
+        // upload buffer data
+        {
+            auto vertex_data = ctx.map_buffer(ui_vertex_buffer);
+            auto index_data = ctx.map_buffer(ui_index_buffer);
+
+            for (auto const& rl : ui_merger.get_render_data().lists)
+            {
+                std::memcpy(vertex_data, rl.vertices.data(), rl.vertices.size_bytes());
+                std::memcpy(index_data, rl.indices.data(), rl.indices.size_bytes());
+
+                vertex_data += rl.vertices.size_bytes();
+                index_data += rl.indices.size_bytes();
+            }
+
+            ctx.unmap_buffer(ui_vertex_buffer);
+            ctx.unmap_buffer(ui_index_buffer);
+        }
 
         auto backbuffer = ctx.acquire_backbuffer(swapchain);
         auto frame = ctx.make_frame();
 
         {
             auto fb = frame.make_framebuffer(backbuffer);
-            auto pass = fb.make_pass(pr::graphics_pass({}, vs_clear, ps_clear));
-            pass.draw(3);
+
+            // clear
+            {
+                auto pass = fb.make_pass(pr::graphics_pass({}, vs_clear, ps_clear));
+                pass.draw(3);
+            }
+
+            // ui
+            {
+                auto pass = fb.make_pass(pso_ui).bind(font_arg, ui_proj);
+                CC_ASSERT(ui_merger.get_render_data().lists.size() == 1 && "more not supported currently");
+                pass.draw(ui_vertex_buffer, ui_index_buffer);
+            }
+
+            if (show_wireframe)
+            {
+                auto pass = fb.make_pass(pso_ui_wf).bind(ui_proj);
+                CC_ASSERT(ui_merger.get_render_data().lists.size() == 1 && "more not supported currently");
+                pass.draw(ui_vertex_buffer, ui_index_buffer);
+            }
         }
 
         frame.present_after_submit(backbuffer, swapchain);
         ctx.submit(cc::move(frame));
+
+        // less latency
+        ctx.flush();
```

Comment:
flushing the GPU shouldn't happen every frame, something is wrong if this really improves latency or performance

Reply:
was there a change relating to that in phi in the last 2 or so months? I remember that without the flush I had at least 200ms latency. now it doesn't change the feel at all